Upstream fullscreening with QuickLook from WebKitAdditions

yoelhawa opened the pull request:
#### 92f2950d8402fa8a96d67dfb222bc4903ea9a44e
<pre>
Upstream fullscreening with QuickLook from WebKitAdditions
<a href="https://bugs.webkit.org/show_bug.cgi?id=277105">https://bugs.webkit.org/show_bug.cgi?id=277105</a>
&lt;<a href="https://rdar.apple.com/132525569">rdar://132525569</a>&gt;

Reviewed by Aditya Keerthi.

Upstream the isSpatial and isPanoramic checks for deciding whether to
delegate fullscreen to QuickLook from WebKitAdditions.

* Source/WebCore/platform/cocoa/PhotosFormatSoftLink.h:
* Source/WebCore/platform/cocoa/PhotosFormatSoftLink.mm:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::isPanoramic const):
(WebCore::ImageDecoderCG::isSpatial const):
(WebCore::ImageDecoderCG::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/SourcesCocoa.txt:
Upstream the methods for determining whether to delegate fullscreen to
QuickLook from WebKitAdditions.

Canonical link: <a href="https://commits.webkit.org/281625@main">https://commits.webkit.org/281625@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8f72a21301ff1d9d9f54a46e2498e9b430e9e98

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/60454 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/39805 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13012 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/64377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/10989 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/62584 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/47479 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11218 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/48930 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/7643 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/62485 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37090 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52369 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/29758 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/33787 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/9592 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/9906 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/55674 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/9882 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66108 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/4387 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/9715 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/56298 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/4418 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52343 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/56454 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/3654 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9090 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/35617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/36699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/37790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/36441 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -700,21 +708,36 @@ bool ImageDecoderCG::canDecodeType(const String& mimeType)
-#include <WebKitAdditions/ImageDecoderCGAdditions.cpp>
-#else
-namespace WebCore {
+    constexpr float panoAspectRatioThreshold = 2.0;
```
`pano` -> `panoramicImage`?

(https://webkit.org/code-style-guidelines/#names-full-words)

pxlcoder commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -43,9 +43,17 @@
+#if ENABLE(QUICKLOOK_FULLSCREEN)
+SOFT_LINK_PRIVATE_FRAMEWORK_FOR_SOURCE(WebCore, PhotosFormats);
+SOFT_LINK_FUNCTION_MAY_FAIL_FOR_SOURCE(WebCore, PhotosFormats, PFMetadataImageSourceIsSpatialMedia, bool, (CGImageSourceRef source), (source));
+#endif
```
Can we create a separate `*SoftLink.h/mm` for this?

`DataDetectorsCoreSoftLink` is an example.

pxlcoder commented on Source/WebCore/platform/cocoa/PhotosFormatSoftLink.h:
```diff
@@ -0,0 +1,28 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import <wtf/SoftLinking.h>
```
```suggestion
#import <wtf/SoftLinking.h>

```

pxlcoder commented on Source/WebCore/platform/cocoa/PhotosFormatSoftLink.h:
```diff
@@ -0,0 +1,28 @@
+
+#import <wtf/SoftLinking.h>
+SOFT_LINK_PRIVATE_FRAMEWORK_FOR_SOURCE(WebCore, PhotosFormats);
+SOFT_LINK_FUNCTION_MAY_FAIL_FOR_SOURCE(WebCore, PhotosFormats, PFMetadataImageSourceIsSpatialMedia, bool, (CGImageSourceRef source), (source));
```
This should use the `_FOR_HEADER` variants, like `SOFT_LINK_FRAMEWORK_FOR_HEADER`.

`_FOR_SOURCE` would then be in a `PhotosFormatSoftLink.mm` file.

pxlcoder commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -43,6 +43,9 @@
+#include <WebCore/platform/cocoa/PhotosFormatSoftLink.h>
+#endif
 
 #include "MediaAccessibilitySoftLink.h"
```
```suggestion
#include "MediaAccessibilitySoftLink.h"
#if ENABLE(QUICKLOOK_FULLSCREEN)
#include "PhotosFormatSoftLink.h"
#endif
```

Commit:
Upstream fullscreening with QuickLook from WebKitAdditions
https://bugs.webkit.org/show_bug.cgi?id=277105
<rdar://132525569>

Reviewed by Aditya Keerthi.

Upstream the isSpatial and isPanoramic checks for deciding whether to
delegate fullscreen to QuickLook from WebKitAdditions.

* Source/WebCore/platform/cocoa/PhotosFormatSoftLink.h:
* Source/WebCore/platform/cocoa/PhotosFormatSoftLink.mm:
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::isPanoramic const):
(WebCore::ImageDecoderCG::isSpatial const):
(WebCore::ImageDecoderCG::shouldUseQuickLookForFullscreen const):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.h:
* Source/WebCore/SourcesCocoa.txt:
Upstream the methods for determining whether to delegate fullscreen to
QuickLook from WebKitAdditions.

Canonical link: https://commits.webkit.org/281625@main
